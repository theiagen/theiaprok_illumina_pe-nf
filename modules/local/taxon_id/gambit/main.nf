process GAMBIT {
    tag "$meta.id"
    label "process_medium"

    // Had to update image for this one
    container "us-docker.pkg.dev/general-theiagen/theiagen/gambit:1.1.0"

    input:
    tuple val(meta), path(assembly)
    path gambit_db_genomes
    path gambit_db_signatures

    output:
    tuple val(meta), path("*_gambit.json"), emit: gambit_report
    tuple val(meta), path("*_gambit_closest.csv"), emit: gambit_closest
    tuple val(meta), path("PREDICTED_TAXON"), emit: predicted_taxon
    tuple val(meta), path("PREDICTED_TAXON_RANK"), emit: predicted_taxon_rank
    tuple val(meta), path("NEXT_TAXON"), emit: next_taxon
    tuple val(meta), path("NEXT_TAXON_RANK"), emit: next_taxon_rank
    tuple val(meta), path("MERLIN_TAG"), emit: merlin_tag
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def report_path = "${prefix}_gambit.json"

    """
    set -euo pipefail

    date | tee DATE
    echo "Running Gambit..."

    # set gambit reference dir; will assume that gambit genomes and signatures will be provided by user in tandem or not at all
    # -s evaluates to TRUE if the file exists and has a size greater than zero
    if [[ -s "${gambit_db_genomes}" ]]; then 
      echo "User gabmit db identified; ${gambit_db_genomes} will be utilized for alignment"
      gambit_db_version="\$(basename -- '${gambit_db_genomes}'); \$(basename -- '${gambit_db_signatures}')"
      gambit_db_dir="\${PWD}/gambit_database"
      mkdir \${gambit_db_dir}
      cp ${gambit_db_genomes} \${gambit_db_dir}
      cp ${gambit_db_signatures} \${gambit_db_dir}
    else
     gambit_db_dir="/gambit-db" 
     gambit_db_version="unmodified from gambit container: ~{docker}"
    fi
    
    # Run Gambit first
    gambit -d \${gambit_db_dir} query -f json -o ${report_path} ${assembly} -c ${task.cpus}

    # Parse the Gambit report
    gambit_report_parser.py ${report_path} ${prefix}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gambit: \$(gambit --version)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_gambit.json
    touch ${prefix}_gambit_closest.csv
    touch PREDICTED_TAXON
    touch PREDICTED_TAXON_RANK
    touch NEXT_TAXON
    touch NEXT_TAXON_RANK
    touch MERLIN_TAG
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gambit: \$(gambit --version)
    END_VERSIONS
    """

}