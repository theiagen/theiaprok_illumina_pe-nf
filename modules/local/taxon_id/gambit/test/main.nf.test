nextflow_process {
    name "Test Process GAMBIT"
    script "../main.nf"
    process "GAMBIT"
    config "./nextflow.config"
    
    test("gambit - prokaryote assembly") {
        setup {
            run("GUNZIP") {
                script "../../../../nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [id:'test_sample'],
                        [file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists: true)]
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = file("https://storage.googleapis.com/jlumpe-gambit/public/databases/refseq-curated/1.0/gambit-refseq-curated-1.0.gdb")
                input[2] = file("https://storage.googleapis.com/jlumpe-gambit/public/databases/refseq-curated/1.0/gambit-refseq-curated-1.0.gs")
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.gambit_report,
                    process.out.gambit_closest,
                    process.out.predicted_taxon,
                    process.out.predicted_taxon_rank,
                    process.out.next_taxon,
                    process.out.next_taxon_rank,
                    process.out.merlin_tag,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("gambit - prokaryote assembly - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists: true)
                ])
                input[1] = file("gs://gambit-databases-rp/2.0.0/gambit-metadata-2.0.0-20240628.gdb")
                input[2] = file("gs://gambit-databases-rp/2.0.0/gambit-signatures-2.0.0-20240628.gs")
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}