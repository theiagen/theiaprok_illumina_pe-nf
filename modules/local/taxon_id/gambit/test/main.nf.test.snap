{
    "gambit - prokaryote assembly - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_gambit.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_gambit_closest.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PREDICTED_TAXON:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PREDICTED_TAXON_RANK:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NEXT_TAXON:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NEXT_TAXON_RANK:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MERLIN_TAG:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    "versions.yml:md5,cc435dec972f7e693963314cf50e6bf3"
                ],
                "gambit_closest": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_gambit_closest.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "gambit_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_gambit.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "merlin_tag": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MERLIN_TAG:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "next_taxon": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NEXT_TAXON:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "next_taxon_rank": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NEXT_TAXON_RANK:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "predicted_taxon": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PREDICTED_TAXON:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "predicted_taxon_rank": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PREDICTED_TAXON_RANK:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cc435dec972f7e693963314cf50e6bf3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T13:40:58.088020487"
    },
    "gambit - prokaryote assembly": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_gambit.json:md5,7f719b2bb09061407f768db62e7fc593"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_gambit_closest.csv:md5,4778d29bd6078b49dbe8bf795f30472b"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "PREDICTED_TAXON:md5,d421f55516b1ad163fa21db9bc0e4cdc"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "PREDICTED_TAXON_RANK:md5,cd3d20b9461987685a810f405a9fcfb2"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "NEXT_TAXON:md5,d4cd0dabcf4caa22ad92fab40844c786"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "NEXT_TAXON_RANK:md5,d4cd0dabcf4caa22ad92fab40844c786"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "MERLIN_TAG:md5,d421f55516b1ad163fa21db9bc0e4cdc"
                ]
            ],
            [
                "versions.yml:md5,cc435dec972f7e693963314cf50e6bf3"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T13:39:09.736676028"
    }
}